File: idealingua/idealingua-runtime-rpc-go/src/main/resources/runtime/go/irt/transport_websocket.go
package irt

import (
	"encoding/json"
	"math/rand"
	"time"
)

const (
	DefaultWriteWaitTime      = 10 * time.Second
	DefaultPongWaitTime       = 60 * time.Second
	DefaultPingPeriod         = (DefaultPongWaitTime * 9) / 10
	DefaultMaxPackageSize     = 1024 * 1024 * 16 // 16 MB
	DefaultReadBufferSize     = 1024 * 8         // 8 KB
	DefaultWriteBufferSize    = 1024 * 8         // 8 KB
	DefaultProcessingThreads  = 2
	DefaultRegistrationBuffer = 2
	DefaultConnectionBuffer   = 5

	MessageRandomSymbols = "abcdefghijklmnopqrstuvwxyz0123456789"
	MessageRandomLength  = 32

	MessageKindFailure        = "?:failure"
	MessageKindRPCRequest     = "rpc:request"
	MessageKindRPCResponse    = "rpc:response"
	MessageKindRPCFailure     = "rpc:failure"
	MessageKindBuzzerRequest  = "buzzer:request"
	MessageKindBuzzerResponse = "buzzer:request"
	MessageKindBuzzerFailure  = "buzzer:request"
	MessageKindStreamS2C      = "stream:s2c"
	MessageKindStreamC2S      = "stream:c2s"
)

type WebSocketMessageBase struct {
	Kind string `json:"kind"`
}

type WebSocketRequestMessage struct {
	Kind    string            `json:"kind"`
	Service string            `json:"service,omitempty"`
	Method  string            `json:"method,omitempty"`
	ID      string            `json:"id"`
	Data    json.RawMessage   `json:"data,omitempty"`
	Headers map[string]string `json:"headers,omitempty"`
}

type WebSocketResponseMessage struct {
	Kind string          `json:"kind"`
	Ref  string          `json:"ref"`
	Data json.RawMessage `json:"data,omitempty"`
}

type WebSocketFailureMessage struct {
	Kind  string `json:"kind"`
	Cause string `json:"cause"`
	Data  string `json:"data"`
}

func RandomMessageID(prefix string) string {
	b := make([]byte, MessageRandomLength)
	for i := range b {
		b[i] = MessageRandomSymbols[rand.Intn(len(MessageRandomSymbols))]
	}
	return prefix + string(b)
}

File: idealingua-v1/idealingua-v1-runtime-rpc-typescript/src/main/resources/runtime/typescript/irt/transport/transport.websocket.ts

export enum WebSocketMessageKind {
    Failure = '?:failure',
    RpcRequest = 'rpc:request',
    RpcResponse = 'rpc:response',
    RpcFailure = 'rpc:failure',
    BuzzerRequest = 'buzzer:request',
    BuzzerResponse = 'buzzer:response',
    BuzzerFailure = 'buzzer:failure',
    StreamS2C = 'stream:s2c',
    StreamC2S = 'stream:c2s'
}

export interface WebSocketMessageBase {
    kind: WebSocketMessageKind
}

// Matches WebSocketMessageKind.RpcRequest, WebSocketMessageKind.BuzzerRequest
export interface WebSocketRequestMessage extends WebSocketMessageBase {
    service?: string
    method?: string
    id: string
    data?: any
    headers?: {[key: string]: string}
}

// Matches WebSocketMessageKind.RpcResponse, WebSocketMessageKind.RpcFailure, 
//      WebSocketMessageKind.BuzzerResponse, WebSocketMessageKind.BuzzerFailure
export interface WebSocketResponseMessage extends WebSocketMessageBase {
    ref: string;
    data: any;
}

// Matches WebSocketMessageKind.Failure
export interface WebSocketFailureMessage extends WebSocketMessageBase {
    data: string;
    cause: string;
}
